cirq/ops/arithmetic_operation.py

```diff
@@ -0,0 +1,248 @@
+    the operation is O(2^n) instead of O(4^n) where n is the number of qubits
+    being acted on, by implementing an `_apply_unitary_` function in terms of
+    the registers and the apply function of the child class. It also handles the
+    boilerplate of implementing the `qubits` and `with_qubits` methods.
+
+    Examples:
+        >>> class Add(cirq.ArithmeticOperation):
+        ...     def __init__(self, target_register, input_register):
+        ...         self.target_register = target_register
+        ...         self.input_register = input_register
+        ...
+        ...     def registers(self):
+        ...         return self.target_register, self.input_register
+        ...
+        ...     def with_registers(self, *new_registers):
+        ...         return Add(*new_registers)
+        ...
+        ...     def apply(self, target_value, input_value):
+        ...         return target_value + input_value
+        >>> cirq.unitary(
+        ...     Add(target_register=cirq.LineQubit.range(2),
+        ...         input_register=1)
+        ... ).astype(np.int32)
+        array([[0, 0, 0, 1],
+               [1, 0, 0, 0],
+               [0, 1, 0, 0],
+               [0, 0, 1, 0]], dtype=int32)
+        >>> c = cirq.Circuit.from_ops(
+        ...    cirq.X(cirq.LineQubit(3)),
+        ...    cirq.X(cirq.LineQubit(2)),
+        ...    cirq.X(cirq.LineQubit(6)),
+        ...    cirq.measure(*cirq.LineQubit.range(4, 8), key='before:in'),
+        ...    cirq.measure(*cirq.LineQubit.range(4), key='before:out'),
+        ...
+        ...    Add(target_register=cirq.LineQubit.range(4),
+        ...        input_register=cirq.LineQubit.range(4, 8)),
+        ...
+        ...    cirq.measure(*cirq.LineQubit.range(4, 8), key='after:in'),
+        ...    cirq.measure(*cirq.LineQubit.range(4), key='after:out'),
+        ... )
+        >>> cirq.sample(c).data
+           before:in  before:out  after:in  after:out
+        0          2           3         2          5
+    """
+
+    @abc.abstractmethod
+    def registers(self) -> Sequence[Union[int, Sequence['cirq.Qid']]]:
+        """The data acted upon by the arithmetic operation.
+
+        Each register in the list can either be a classical constant (an `int`),
+        or else a list of qubits/qudits (a `List[cirq.Qid]`). Registers that
+        are set to a classical constant must not be mutated by the arithmetic
+        operation (their value must remain fixed when passed to `apply`).
+
+        Registers are big endian. The first qubit is the most significant, the
+        last qubit is the 1s qubit, the before last qubit is the 2s qubit, etc.
+
+        Returns:
+            A list of constants and qubit groups that the operation will act
+            upon.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def with_registers(self, *new_registers: Union[int, Sequence['cirq.Qid']]
```

Comment:
why would one use this instead of just calling the constructor again?

Reply:
Child classes often have parameters other than the registers in their constructor. For example: a boolean switch determining whether to increment or decrement, or metadata about the source of the gate. They may also follow conventions such as enforcing keywords.